From: Utility Muffin Research Kitchen <umrk@users.noreply.github.com>
Subject: [PATCH] Make retro_arena dependencies cross-build clean

Use the target sysroot's EGL/GLES, PNG, and zlib, pass the toolchain into
libchdr, pin the downloaded JSON headers, retain the shared Cortex-A55 flags,
and replace Boost filesystem existence checks with POSIX access().
---
 yabause/CMake/Packages/external_libchdr.cmake | 11 +++-
 yabause/src/CMakeLists.txt                     | 32 +++++++++---
 yabause/src/retro_arena/CMakeLists.txt         | 25 +++++-----
 yabause/src/retro_arena/InputManager.cpp       |  9 ++--
 yabause/src/retro_arena/main.cpp               |  2 +-
 5 files changed, 50 insertions(+), 29 deletions(-)

diff --git a/yabause/CMake/Packages/external_libchdr.cmake b/yabause/CMake/Packages/external_libchdr.cmake
--- a/yabause/CMake/Packages/external_libchdr.cmake
+++ b/yabause/CMake/Packages/external_libchdr.cmake
@@ -24,7 +24,16 @@ elseif(IOS)
   )
 
 else()
-  set(ADDITIONAL_CMAKE_ARGS "")
+  if(CMAKE_CROSSCOMPILING)
+    get_filename_component(TOOL_CHAIN_ABSOLUTE_PATH "${CMAKE_TOOLCHAIN_FILE}"
+                           REALPATH BASE_DIR "${CMAKE_BINARY_DIR}")
+    set(ADDITIONAL_CMAKE_ARGS
+      -DCMAKE_TOOLCHAIN_FILE=${TOOL_CHAIN_ABSOLUTE_PATH}
+      -DCMAKE_SKIP_RPATH=ON
+      -DCMAKE_POLICY_VERSION_MINIMUM=3.5)
+  else()
+    set(ADDITIONAL_CMAKE_ARGS "")
+  endif()
 endif()
 
 ExternalProject_Add(
diff --git a/yabause/src/CMakeLists.txt b/yabause/src/CMakeLists.txt
--- a/yabause/src/CMakeLists.txt
+++ b/yabause/src/CMakeLists.txt
@@ -337,12 +337,26 @@ endif ()
 # OpenGL
 option(YAB_WANT_OPENGL "use OpenGL for video output (most ports require it)" ON)
 if (YAB_WANT_OPENGL AND (YAB_RGB STREQUAL ""))
-	include(FindOpenGL)
+	if(USE_EGL)
+		find_path(EGL_INCLUDE_DIR EGL/egl.h)
+		find_library(EGL_LIBRARY EGL)
+		find_path(OPENGLES3_INCLUDE_DIR GLES3/gl3.h)
+		find_library(OPENGLES2_LIBRARY GLESv2)
+		if(NOT EGL_INCLUDE_DIR OR NOT EGL_LIBRARY OR
+		   NOT OPENGLES3_INCLUDE_DIR OR NOT OPENGLES2_LIBRARY)
+			message(FATAL_ERROR "EGL and OpenGL ES 3 are required for USE_EGL")
+		endif()
+		set(OPENGL_FOUND TRUE)
+		include_directories(${EGL_INCLUDE_DIR} ${OPENGLES3_INCLUDE_DIR})
+	else()
+		include(FindOpenGL)
+	endif()
 	if (OPENGL_FOUND )
 		add_definitions(-DHAVE_LIBGL=1)
 		if( USE_EGL )
 			add_definitions(-D_OGLES3_)
-			set(YABAUSE_LIBRARIES ${YABAUSE_LIBRARIES} EGL GLESv2)
+			set(YABAUSE_LIBRARIES ${YABAUSE_LIBRARIES}
+				${EGL_LIBRARY} ${OPENGLES2_LIBRARY})
 			add_definitions(-DYAB_ASYNC_RENDERING)
 		else()
 			add_definitions(-D_OGL3_)
@@ -985,8 +999,8 @@ if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "armv7-a")
 endif()
 if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "aarch64")
 		add_definitions(-DAARCH64)
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -g -std=c99 -march=armv8-a")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g -march=armv8-a -std=c++17 ")
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu99")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17")
 endif()
 
 # Warnings defined to know when we're breaking compilation with MSVC
@@ -1037,10 +1051,10 @@ endif(YAB_WANT_VULKAN)
 #-----------------------------------------------------------------------
 # libCHD
 include( ${CMAKE_SOURCE_DIR}/CMake/Packages/external_libchdr.cmake )
-include( ${CMAKE_SOURCE_DIR}/CMake/Packages/external_zlib.cmake )
-#include_directories(${zlib_INCLUDE_DIR})
+find_package(ZLIB REQUIRED)
+set(zlib_INCLUDE_DIR ${ZLIB_INCLUDE_DIRS})
+set(zlib_STATIC_LIBRARIES ${ZLIB_LIBRARIES})
 add_dependencies(yabause libchdr)
-#add_dependencies(yabause zlib)
 
 if(WIN32)
 	# Windows need to build zlib
diff --git a/yabause/src/retro_arena/CMakeLists.txt b/yabause/src/retro_arena/CMakeLists.txt
--- a/yabause/src/retro_arena/CMakeLists.txt
+++ b/yabause/src/retro_arena/CMakeLists.txt
@@ -10,8 +10,6 @@ find_package(SDL2 REQUIRED)
 find_package(OpenGLES)
 find_package(PkgConfig REQUIRED)
 
-find_package(Boost REQUIRED COMPONENTS system filesystem date_time locale)
-
 if (NOT OPENGLES2_FOUND)
 	return()
 endif (NOT OPENGLES2_FOUND)
@@ -21,8 +19,8 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
 add_subdirectory(pugixml)
 add_subdirectory(nanogui-sdl)
 
-#include( ../glfw/external_zlib.cmake )
-include( ${CMAKE_SOURCE_DIR}/CMake/Packages/external_libpng.cmake )
+find_package(PNG REQUIRED)
+find_package(ZLIB REQUIRED)
 
 #------------------------------------------------------------------------------
 # nlohmann/json
@@ -31,15 +29,18 @@ include( ${CMAKE_SOURCE_DIR}/CMake/Packages/external_libpng.cmake )
 ExternalProject_Add(
      Json
      PREFIX ${CMAKE_CURRENT_BINARY_DIR}/Json
-     URL https://github.com/nlohmann/json/archive/v3.2.0.zip
+     GIT_REPOSITORY https://github.com/nlohmann/json.git
+     GIT_TAG 359f98d14065bf4e53eeb274f5987fd08f16e5bf
      #disable following
      CONFIGURE_COMMAND "" BUILD_COMMAND "" INSTALL_DIR "" INSTALL_COMMAND ""
      )
 # set include directory
 set(EXTERNAL_JSON_INCLUDE_DIR "${CMAKE_CURRENT_BINARY_DIR}/Json/src/Json/single_include")
 
-set(PORT_INCLUDE_DIRS ${EXTERNAL_JSON_INCLUDE_DIR} ${SDL2_INCLUDE_DIR} ${png_INCLUDE_DIR} ${zlib_INCLUDE_DIR} )
-set(PORT_LIBRARIES ${SDL2_LIBRARY} ${OPENGLES2_LIBRARIES} ${png_STATIC_LIBRARIES} ${zlib_STATIC_LIBRARIES} )
+set(PORT_INCLUDE_DIRS ${EXTERNAL_JSON_INCLUDE_DIR} ${SDL2_INCLUDE_DIR}
+    ${PNG_INCLUDE_DIRS} ${ZLIB_INCLUDE_DIRS})
+set(PORT_LIBRARIES ${SDL2_LIBRARY} ${OPENGLES2_LIBRARIES}
+    ${PNG_LIBRARIES} ${ZLIB_LIBRARIES})
 
 message("Nlohmann/json dir =" ${EXTERNAL_JSON_INCLUDE_DIR})
 
@@ -57,12 +58,14 @@ set(yabause_ra_SOURCES
     ../about.c
 ) 
 link_directories( ${link_directories} ${LIBCHDR_LIB_DIR} )
-include_directories( ${PORT_INCLUDE_DIRS} ${Boost_INCLUDE_DIR} "nanogui-sdl/include" "nanogui-sdl/ext/eigen" )
+include_directories( ${PORT_INCLUDE_DIRS} "nanogui-sdl/include" "nanogui-sdl/ext/eigen" )
 add_executable(yabause-retro-arena ${yabause_ra_SOURCES})
-target_link_libraries(yabause-retro-arena yabause  ${Boost_LIBRARIES} pugixml ${YABAUSE_LIBRARIES} ${PORT_LIBRARIES} nanogui stdc++fs  ${LIBCHDR_LIBRARIES} )
+target_compile_definitions(yabause PRIVATE __RETORO_ARENA__)
+target_link_libraries(yabause-retro-arena yabause pugixml ${YABAUSE_LIBRARIES}
+    ${PORT_LIBRARIES} nanogui stdc++fs ${LIBCHDR_LIBRARIES})
 target_link_libraries( yabause-retro-arena ${LIBCHDR_LIBRARIES} )
-target_link_libraries( yabause-retro-arena ${zlib_STATIC_LIBRARIES} )
-add_dependencies(yabause-retro-arena Json png zlib libchd )
+target_link_libraries( yabause-retro-arena ${ZLIB_LIBRARIES} )
+add_dependencies(yabause-retro-arena Json libchdr)
 yab_port_success(yabause-retro-arena)
 set_target_properties(yabause-retro-arena PROPERTIES OUTPUT_NAME yabasanshiro)
 
diff --git a/yabause/src/retro_arena/InputManager.cpp b/yabause/src/retro_arena/InputManager.cpp
--- a/yabause/src/retro_arena/InputManager.cpp
+++ b/yabause/src/retro_arena/InputManager.cpp
@@ -24,7 +24,6 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
 //#include "Window.h"
 //#include "Log.h"
 #include "pugixml/pugixml.hpp"
-#include <boost/filesystem.hpp>
 //#include "platform.h"
 #include <iostream>
 
@@ -78,8 +77,6 @@ static const unsigned int SDL_HAT_VALUES_NUM = sizeof(SDL_HAT_VALUES) / sizeof(S
 // 4. Joystick GUID - this is some squashed version of joystick vendor, version, and a bunch of other device-specific things.
 //    It should remain the same across runs of the program/system restarts/device reordering and is what I use to identify which joystick to load.
 
-namespace fs = boost::filesystem;
-
 InputManager* InputManager::mInstance = NULL;
 void genJoyString( string & out, SDL_JoystickID id, const string & name, const string & guid );
 
@@ -608,7 +605,7 @@ void InputManager::init( const std::string & fname )
 
   json j;
   config_fname_ = fname;
-  if(!fs::exists(fname)){
+  if(access(fname.c_str(), F_OK) != 0){
     if( convertFromEmustationFile(fname) != 0 ){
       return;
     }
@@ -1119,7 +1116,7 @@ bool InputManager::parseEvent(const SDL_Event& ev)
 bool InputManager::loadInputConfig(InputConfig* config)
 {
   std::string path = getConfigPath();
-  if(!fs::exists(path)){
+  if(access(path.c_str(), F_OK) != 0){
     PADLOG("loadInputConfig %s is not found!\n", path.c_str() );
     return false;
   }
@@ -1176,7 +1173,7 @@ void InputManager::writeDeviceConfig(InputConfig* config)
 
   pugi::xml_document doc;
 
-  if(fs::exists(path))
+  if(access(path.c_str(), F_OK) == 0)
   {
     // merge files
     pugi::xml_parse_result result = doc.load_file(path.c_str());
diff --git a/yabause/src/retro_arena/main.cpp b/yabause/src/retro_arena/main.cpp
--- a/yabause/src/retro_arena/main.cpp
+++ b/yabause/src/retro_arena/main.cpp
@@ -389,7 +389,7 @@ int main(int argc, char** argv)
 	//SDL_GL_SetAttribute(SDL_GL_BLUE_SIZE, 8);
 
 		SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK,  SDL_GL_CONTEXT_PROFILE_ES);
-		SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 2);
+		SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
 		SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 0);
 
 		SDL_GL_SetAttribute(SDL_GL_RED_SIZE,           8);
